Sources/SourceKitLSP/IndexOutOfDateChecker.swift

```diff
@@ -0,0 +1,101 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import IndexStoreDB
+import LSPLogging
+
+/// Helper class to check if symbols from the index are up-to-date or if the source file has been modified after it was
+/// indexed.
+///
+/// The checker caches mod dates of source files. It should thus not be long lived. Its intended lifespan is the
+/// evaluation of a single request.
+struct IndexOutOfDateChecker {
+  /// The last modification time of a file. Can also represent the fact that the file does not exist.
+  private enum ModificationTime {
+    case fileDoesNotExist
+    case date(Date)
+  }
+
+  private enum Error: Swift.Error, CustomStringConvertible {
+    case fileAttributesDontHaveModificationDate
+
+    var description: String {
+      switch self {
+      case .fileAttributesDontHaveModificationDate:
+        return "File attributes don't contain a modification date"
+      }
+    }
+  }
+
+  /// File paths to modification times that have already been computed.
+  private var modTimeCache: [String: ModificationTime] = [:]
+
+  private func modificationDateUncached(of path: String) throws -> ModificationTime {
+    do {
+      let attributes = try FileManager.default.attributesOfItem(atPath: path)
+      guard let modificationDate = attributes[FileAttributeKey.modificationDate] as? Date else {
+        throw Error.fileAttributesDontHaveModificationDate
+      }
+      return .date(modificationDate)
+    } catch let error as NSError where error.domain == NSCocoaErrorDomain && error.code == NSFileReadNoSuchFileError {
+      return .fileDoesNotExist
+    }
+  }
+
+  private mutating func modificationDate(of path: String) throws -> ModificationTime {
+    if let cached = modTimeCache[path] {
+      return cached
+    }
+    let modTime = try modificationDateUncached(of: path)
+    modTimeCache[path] = modTime
+    return modTime
+  }
+
+  /// Returns `true` if the source file for the given symbol location exists and has not been modified after it has been
+  /// indexed.
+  mutating func isUpToDate(_ symbolLocation: SymbolLocation) -> Bool {
+    do {
+      let sourceFileModificationDate = try modificationDate(of: symbolLocation.path)
+      switch sourceFileModificationDate {
+      case .fileDoesNotExist:
+        return false
+      case .date(let sourceFileModificationDate):
+        return sourceFileModificationDate <= symbolLocation.timestamp
+      }
+    } catch {
+      logger.fault("Unable to determine if SymbolLocation is up-to-date: \(error.forLogging)")
+      return true
```

Comment:
The only error at the moment is no mtime in attributes. Seems like we should just assume it isn't up to date in that case? I doubt it really matters though.

Reply:
My thinking was that if sourcekit-lsp is running on some platform that for some reason doesn’t have modification time or if Foundation doesn’t know how to extract it (not sure if those exist), it’s better to assume all files are up-to-date than to completely ignore index results. But as you said, I don’t think it really makes a difference.

Reply:
Fair enough, that's fine with me 👍